src/metabase/actions/actions.clj

```diff
@@ -177,6 +176,32 @@
   [action-or-id]
   (check-actions-enabled-for-database! (api/check-404 (database-for-action action-or-id))))
 
+(defn- row-update-event [{:keys [before after]}]
+  (case [(some? before) (some? after)]
+    [false true]  :event/rows.created
+    [true  true]  :event/rows.updated
+    [true  false] :event/rows.deleted
+    ;; should not happen
+    [false false] ::no-op))
+
+(defn- publish-effect-events! [{:keys [user-id effects]}]
+  (doseq [[event payloads] (->> effects
+                                (filter (comp #{:effect/row.modified} first))
+                                (map second)
+                                (group-by row-update-event)
+                                (remove (comp #{::no-op} key)))]
+    ;; TODO will fix requiring-resolve when we remove all this table-action specific stuff, which has NO place here
+    #_{:clj-kondo/ignore [:metabase/modules]}
+    (doseq [[table-id payloads] (group-by :table-id payloads)
+            :let [pk-fields   ((requiring-resolve 'metabase-enterprise.data-editing.data-editing/select-table-pk-fields) table-id)
+                  row-changes (for [{:keys [before after]} payloads]
+                                {:pk     ((requiring-resolve 'metabase-enterprise.data-editing.data-editing/get-row-pks) pk-fields after)
+                                 :before before
+                                 :after  after})]]
+      (events/publish-event! event {:actor_id    user-id
+                                    :row-changes row-changes
```

Comment:
```suggestion
                                    :row_changes row-changes
```
it's not directly used in the template but I think we should have a consistent naming here too.

Reply:
So the convention I think we should follow is to always convert to snake case at the last possible moment, when it crosses into the frontend or the database. Everywhere else in the app should use hyphens. I believe this is how most of our repo is structured.

Reply:
You mean because the frontend can *write* conditional expressions? I will never evaluate them, right? So technically it's never working with the actual payload values, just talking about the same schema?

I guess we could adopt the convention of using only underscores in event payloads, but we really need to stop mixing them, it makes things a lot more error prone and burdensome to get right.